modules/fileout_netcdf/FONcArrayStructureField.cc

```diff
@@ -0,0 +1,366 @@
+    if(b_data_type == libdap::dods_array_c)
+        supported_atomic_data_type = b->var()->type();
+    if (!is_simple_type(supported_atomic_data_type) || supported_atomic_data_type == dods_str_c
+        || supported_atomic_data_type == dods_url_c
+        || supported_atomic_data_type == dods_enum_c || supported_atomic_data_type ==dods_opaque_c) {
+        string s = "File out netcdf, only support one-layer of simple int/float fields inside an array of structure.";
+        throw BESInternalError(s, __FILE__, __LINE__);
+    }
+
+    d_a = a;
+    d_varname = b->name();
+    if (b_data_type == libdap::dods_array_c) {
+        d_array_type = FONcUtils::get_nc_type(b->var(), isNetCDF4_ENHANCED());
+        d_array_type_size = (b->var()->width_ll())/(b->var()->length_ll());
+    }
+    else {
+        d_array_type = FONcUtils::get_nc_type(b, isNetCDF4_ENHANCED());
+        d_array_type_size = (b->width_ll())/(b->length_ll());
+    }
+    // Need to retrieve dimension information here.
+    // In this version, we only try to get the dimension size right.
+    Array::Dim_iter di = d_a->dim_begin();
+    Array::Dim_iter de = d_a->dim_end();
+    for (; di != de; di++) {
+        int64_t size = d_a->dimension_size_ll(di, true);
+        struct_dim_sizes.push_back(size);
+        total_nelements *= size;
+        FONcDim *use_dim = find_sdim(d_a->dimension_name(di), size);
+        struct_dims.push_back(use_dim);
+    }
+    if (b_data_type == libdap::dods_array_c) {
+        auto db_a = dynamic_cast<Array *>(b);
+        if (!db_a){
+            string s = "File out netcdf, FONcArrayStructField was passed a variable that is not a DAP Array";
+            throw BESInternalError(s, __FILE__, __LINE__);
+        }
+        Array::Dim_iter b_di = db_a->dim_begin();
+        Array::Dim_iter b_de = db_a->dim_end();
+        for (; b_di != b_de; b_di++) {
+            int64_t size = d_a->dimension_size_ll(b_di, true);
+            struct_dim_sizes.push_back(size);
+            total_nelements *= size;
+            field_nelements *= size;
+            FONcDim *use_dim = find_sdim(db_a->dimension_name(b_di), size);
+            struct_dims.push_back(use_dim);
+        }
+    }
+
+}
+
+void FONcArrayStructureField::convert(vector<string> embed, bool _dap4, bool is_dap4_group){
+    var_name = FONcUtils::gen_name(embed,d_varname, d_orig_varname);
+
+}
+
+/** @brief define the DAP array of structure field  in the netcdf file
+ *
+ * @param ncid The id of the NetCDF file
+ * @throws BESInternalError if there is a problem. 
+ */
+void
+FONcArrayStructureField::define( int ncid )
+{
+    BESDEBUG("fonc", "FONcArray::define() - defining array '" << d_varname << "'" << endl);
+
+    if (!d_defined) {
+
+        BESDEBUG("fonc", "FONcArray::define() - defining array of structure field: " << d_varname << "'" << endl);
+        auto i = struct_dims.begin();
+        auto e = struct_dims.end();
+        for (; i != e; i++) {
```

Comment:
You could use the range-based for here, too.

Reply:
Yes, change the code.